File: Source/Resources/Calendar.js
//NOTE: alert() statements are available for debugging purposes. You can uncomment the statements to show dialogs when each method is hit.

(function ($) {
    if (typeof SourceCodeANZ === "undefined" || SourceCodeANZ == null) SourceCodeANZ = {};
    if (typeof SourceCodeANZ.Forms === "undefined" || SourceCodeANZ.Forms == null) SourceCodeANZ.Forms = {};
    if (typeof SourceCodeANZ.Forms.Controls === "undefined" || SourceCodeANZ.Forms.Controls == null) SourceCodeANZ.Forms.Controls = {};
    
    SourceCodeANZ.Forms.Controls.Calendar = {

        //internal method used to get a handle on the control instance
        _getInstance: function (id) {
            var control = jQuery('#' + id);
            if (control.length == 0) {
                throw 'SourceCodeANZ.Forms.Controls.Calendar \'' + id + '\' not found';
            } else {
                return control[0];
            }
        },

        execute: function (objInfo) {
            var method = objInfo.methodName;
            if (method) {
                var _calendar = $("#" + objInfo.CurrentControlID);
                switch (method) {
                    case "Refresh":
                        _calendar[0].attributes['loaddataoninit'].value = "True";
                        _calendar.fullCalendar('refetchEvents');
                        break;
                    case "GoToDate":
                        var parameters = objInfo.methodParameters;
                        var targetdate = new Date(parameters["date"]);
                        _calendar.fullCalendar('gotoDate', targetdate);
                        break;
                }
            }
        },


        
        getProperty: function (objInfo) {
            var instance = SourceCodeANZ.Forms.Controls.Calendar._getInstance(objInfo.CurrentControlId);
            return instance.attributes[objInfo.property].value;
        },

        setProperty: function (objInfo) {
            //debugger;
            switch (objInfo.property.toLowerCase()) {
                case "isvisible":
                    SourceCodeANZ.Forms.Controls.Calendar.setIsVisible(objInfo);
                    break;
                case "isenabled":
                    SourceCodeANZ.Forms.Controls.Calendar.setIsEnabled(objInfo);
                    break;
                case "isreadonly":
                    SourceCodeANZ.Forms.Controls.Calendar.SetControlIsReadOnly(objInfo);
                    break;
                case "locale":
                    $('#' + objInfo.CurrentControlId).fullCalendar('option', 'locale', objInfo.Value);
                    break;
                default:
                    //$('#' + objInfo.CurrentControlId).data(objInfo.property).value = objInfo.Value;
                    var instance = SourceCodeANZ.Forms.Controls.Calendar._getInstance(objInfo.CurrentControlId);
                    instance.attributes[objInfo.property].value = objInfo.Value;
            }
        },


        //helper method to set visibility
        setIsVisible: function (objInfo) {
            value = (objInfo.Value === true || objInfo.Value == 'true');
            this._isVisible = value;
            var displayValue = (value === false) ? "none" : "block";
            var instance = SourceCodeANZ.Forms.Controls.Calendar._getInstance(objInfo.CurrentControlId);
            instance.style.display = displayValue;
        },

       
        SetControlIsReadOnly: function (objInfo) {
            //debugger;
            value = (objInfo.Value === true || objInfo.Value == 'true');
            var instance = SourceCodeANZ.Forms.Controls.Calendar._getInstance(objInfo.CurrentControlId);
            instance.attributes["isreadonly"].value = value;
        },


        //helper method to set control "enabled" state
        setIsEnabled: function (objInfo) {
            value = (objInfo.Value === true || objInfo.Value == 'true');
            this._isEnabled = value;
            var instance = SourceCodeANZ.Forms.Controls.Calendar._getInstance(objInfo.CurrentControlId);
            instance.readOnly = !value;
        }
    };


    $.widget('sfc.calendar1',
{
    _calendar: null,

    _create: function () {
        var date = new Date();
        var d = date.getDate();
        var m = date.getMonth();
        var y = date.getFullYear();
        _calendar = jQuery("#" + this.element.attr('id'));
        var customneweventui = _calendar.attr('customneweventui');
        var self = this;
        var selfControl = self.element[0];
        _calendar.fullCalendar({
            locale: this.options.Locale,
            header: {
                left: 'prevYear prev,today,next nextYear',
                center: 'title',
                right: selfControl.attributes["header_right"].value
            },
            defaultView: selfControl.attributes["header_right"].value.split(",")[0],
            views: {
                month: {
                    titleFormat: selfControl.attributes["monthviewtitleformat"].value,
                    columnFormat: selfControl.attributes["monthviewcolformat"].value
                },
                week: {
                    titleFormat: selfControl.attributes["weekviewtitleformat"].value,
                    columnFormat: selfControl.attributes["weekviewcolformat"].value
                },
                day: {
                    titleFormat: selfControl.attributes["dayviewtitleformat"].value,
                    columnFormat: selfControl.attributes["dayviewcolformat"].value
                },
                basicDay: {
                    titleFormat: selfControl.attributes["dayviewtitleformat"].value
                },
                basicWeek: {
                    titleFormat: selfControl.attributes["weekviewtitleformat"].value
                }
            },
            theme: false,
            selectable: true,
            selectHelper: true,
            lazyFetching: false,
            select: function (start, end) {
                if (!end.hasTime())
                    end.subtract(1, 'days');
                if (customneweventui == "False" && !(selfControl.attributes["isreadonly"].value == "true")) {
                    var title = prompt('Event Title:');
                    if (title) {
                        $.ajax({
                            url: 'CalendarAjaxHandler.handler',
                            cache: false,
                            type: 'POST',
                            dataType: "json",
                            data: {
                                command: "Create",
                                utcoffset: moment().utcOffset() / 60,
                                idfieldname: self.options["SMOValueProperty"],
                                data_source: self.options["SmartObject"],
                                title: title,
                                startAsDate: start.format("YYYY-MM-DD HH:mm"),
                                endAsDate: end.format("YYYY-MM-DD HH:mm"),
                                allDayAsBool: (!(start.hasTime() || end.hasTime())).toString(),
                                color: "#d9e57f",
                                SmartObjectName: self.options["SmartObject"],
                                SmartObjectMethod: self.options["SmartObjectMethod"],
                                SmartObjectValueProp: self.options["SMOValueProperty"],
                                SmartObjectDisplayTemplate: self.options["SMODisplayTemplate"],
                                EventTitleField: self.options["EventTitleField"],
                                EventStartField: self.options["EventStartField"],
                                EventEndField: self.options["EventEndField"],
                                EventAllDayField: self.options["EventAllDayField"],
                                EventEditableField: self.options["EventEditableField"],
                                EventSizeableField: self.options["EventSizeableField"],
                                EventColorField: self.options["EventColorField"],
                                SOCreateMethod: selfControl.attributes["socreatemethod"].value,
                                SOSaveMethod: selfControl.attributes["sosavemethod"].value
                            },
                            success: function (text) {
                                _calendar.fullCalendar('renderEvent',
                               {
                                   id: text[0].id,
                                   title: title,
                                   start: start,
                                   end: end,
                                   allDay: (!(start.hasTime() || end.hasTime())).toString(),
                                   eventdatasource: self.options["SmartObject"],
                                   eventdatasourceid: self.options["SMOValueProperty"]
                               },
                               false
                                );
                                selfControl.attributes["value"].value = text[0].id;
                                selfControl.attributes["eventtitle"].value = title;
                                selfControl.attributes["eventstart"].value = start.format("YYYY-MM-DD HH:mm");
                                selfControl.attributes["eventend"].value = end.format("YYYY-MM-DD HH:mm");
                                selfControl.attributes["eventallday"].value = !(start.hasTime() || end.hasTime());
                                raiseEvent(selfControl.id, 'Control', 'OnChange');
                            },
                            error: function (text) {
                                alert('there was an error while trying to execute event create');
                            }
                        });
                    }
                    _calendar.fullCalendar('unselect');
                } else {
                    selfControl.attributes["value"].value = "0";
                    selfControl.attributes["eventtitle"].value = "";
                    selfControl.attributes["eventcolor"].value = "";
                    selfControl.attributes["eventstart"].value = start.format("YYYY-MM-DD HH:mm");
                    selfControl.attributes["eventend"].value = end.format("YYYY-MM-DD HH:mm");
                    selfControl.attributes["eventallday"].value = (!(start.hasTime() || end.hasTime())).toString();
                    raiseEvent(selfControl.id, 'Control', 'OnClick');
                }
            },
            editable: true,
            eventClick: function(calEvent, jsEvent, view) {
                selfControl.attributes["value"].value = calEvent.id;
                selfControl.attributes["eventtitle"].value = calEvent.title;
                selfControl.attributes["eventstart"].value = calEvent.start.format("YYYY-MM-DD HH:mm");
                if (calEvent.end)
                    selfControl.attributes["eventend"].value = calEvent.end.format("YYYY-MM-DD HH:mm");
                else
                    selfControl.attributes["eventend"].value = calEvent.start.format("YYYY-MM-DD HH:mm");
                selfControl.attributes["eventallday"].value = calEvent.allDay;
                selfControl.attributes["eventeditable"].value = calEvent.editable;
                selfControl.attributes["eventsizeable"].value = calEvent.durationEditable;
                selfControl.attributes["eventcolor"].value = calEvent.color;
                selfControl.attributes["eventtype"].value = calEvent.eventtype;
                raiseEvent(selfControl.id, 'Control', 'OnClick');
            },
            eventAfterAllRender: function (view) {
                //This will force SmartForms Modal Pop-ups to correctly disable everything behind the pop-up:
                $(".fc-button").css("z-index", "0");
                $(".fc-content").css("z-index", "0");
            },
            
            events: function (start, end, timezone , callback) {
                   $.ajax({
                       url: 'CalendarAjaxHandler.handler',
                       cache: false,
                       dataType: "json",
                       data: {
                           start: start.format("YYYY-MM-DD HH:mm"),
                           end: end.format("YYYY-MM-DD HH:mm"),
                           command: "GetList",
                           utcoffset: moment().utcOffset() / 60,
                           useserverzone: self.options["UseServerZone"],
                           ControlIsReadOnly: false,
                           SmartObjectName: self.options["SmartObject"],
                           SmartObjectMethod: self.options["SmartObjectMethod"],
                           SmartObjectValueProp: self.options["SMOValueProperty"],
                           SmartObjectDisplayTemplate: self.options["SMODisplayTemplate"],
                           EventTitleField: self.options["EventTitleField"],
                           EventStartField: self.options["EventStartField"],
                           EventEndField: self.options["EventEndField"],
                           EventAllDayField: self.options["EventAllDayField"],
                           EventEditableField: self.options["EventEditableField"],
                           EventSizeableField: self.options["EventSizeableField"],
                           EventColorField: self.options["EventColorField"],
                           EventTypeField: self.options["EventTypeField"],
                           EventDataSourceField: self.options["EventDataSourceField"],
                           EventDataSourceIDfield: self.options["EventDataSourceIDfield"],
                           FilterSpec: selfControl.attributes["filterspec"].value,
                           MethodParameters: selfControl.attributes["methodparameters"].value,
                           LoadDataOnInt: selfControl.attributes["loaddataoninit"].value
                       },
                       success: function (data, text, xhr) {
                           //console.log("...SourceCodeANZ.Forms.Controls.Calendar...Data Loaded");
                           callback(data);
                       },
                       error: function (xhr, status, error) {
                           alert('There was an error while fetching events!');
                       }
                   });
               }
            ,
            eventResize: function (event, dayDelta, revertFunc) {
                if (!(selfControl.attributes["isreadonly"].value == "true")) {
                    var newEnd = moment(event.end);
                    if (!newEnd.hasTime())
                        newEnd.subtract(1, 'days');
                    $.ajax({
                        url: 'CalendarAjaxHandler.handler',
                        cache: false,
                        type: 'POST',
                        dataType: "json",
                        data: {
                            command: "Update",
                            utcoffset: moment().utcOffset() / 60,
                            id: event.id,
                            title: event.title,
                            color: event.color,
                            idfieldname: event.eventdatasourceid,
                            data_source: event.eventdatasource,
                            startAsDate: event.start.format("YYYY-MM-DD HH:mm"),
                            endAsDate: newEnd.format("YYYY-MM-DD HH:mm"),
                            allDayAsBool: event.allDay.toString(),
                            SmartObjectName: self.options["SmartObject"],
                            SmartObjectMethod: self.options["SmartObjectMethod"],
                            SmartObjectValueProp: self.options["SMOValueProperty"],
                            SmartObjectDisplayTemplate: self.options["SMODisplayTemplate"],
                            EventTitleField: self.options["EventTitleField"],
                            EventStartField: self.options["EventStartField"],
                            EventEndField: self.options["EventEndField"],
                            EventAllDayField: self.options["EventAllDayField"],
                            EventEditableField: self.options["EventEditableField"],
                            EventSizeableField: self.options["EventSizeableField"],
                            EventDataSourceField: self.options["EventDataSourceField"],
                            EventDataSourceIDfield: self.options["EventDataSourceIDfield"],
                            SOCreateMethod: selfControl.attributes["socreatemethod"].value,
                            SOSaveMethod: selfControl.attributes["sosavemethod"].value
                        },
                        success: function (data, text, xhr) {
                            selfControl.attributes["value"].value = data[0].id;
                            selfControl.attributes["eventtitle"].value = data[0].title;
                            selfControl.attributes["eventstart"].value = data[0].start;
                            selfControl.attributes["eventend"].value = data[0].end;
                            selfControl.attributes["eventallday"].value = data[0].allDay;
                            raiseEvent(selfControl.id, 'Control', 'OnChange');
                        },
                        error: function (text) {
                            alert('there was an error while trying to execute event resize');
                            revertFunc();
                        }
                    });

                }
                else revertFunc();
            },
            eventDrop: function (event, dayDelta, revertFunc) {
                if (!(selfControl.attributes["isreadonly"].value == "true")) {
                    var newEnd = moment(event.end);
                    if (!newEnd.hasTime())
                        newEnd.subtract(1, 'days');
                    var frmStart = event.start.format("YYYY-MM-DD HH:mm");
                    var frmEnd;
                    if (event.end)
                        frmEnd = newEnd.format("YYYY-MM-DD HH:mm");
                    else
                        frmEnd = frmStart;
                    $.ajax({
                        url: 'CalendarAjaxHandler.handler',
                        cache: false,
                        type: 'POST',
                        dataType: "json",
                        data: {
                            command: "Update",
                            utcoffset: moment().utcOffset() / 60,
                            id: event.id,
                            title: event.title,
                            color: event.color,
                            idfieldname: event.eventdatasourceid,
                            data_source: event.eventdatasource,
                            startAsDate: frmStart,
                            endAsDate: frmEnd,
                            SmartObjectName: self.options["SmartObject"],
                            SmartObjectMethod: self.options["SmartObjectMethod"],
                            SmartObjectValueProp: self.options["SMOValueProperty"],
                            SmartObjectDisplayTemplate: self.options["SMODisplayTemplate"],
                            EventTitleField: self.options["EventTitleField"],
                            EventStartField: self.options["EventStartField"],
                            EventEndField: self.options["EventEndField"],
                            EventAllDayField: self.options["EventAllDayField"],
                            EventEditableField: self.options["EventEditableField"],
                            EventSizeableField: self.options["EventSizeableField"],
                            EventDataSourceField: self.options["EventDataSourceField"],
                            EventDataSourceIDfield: self.options["EventDataSourceIDfield"],
                            SOCreateMethod: selfControl.attributes["socreatemethod"].value,
                            SOSaveMethod: selfControl.attributes["sosavemethod"].value
                        },
                        success: function (data, text, xhr) {
                            selfControl.attributes["value"].value = data[0].id;
                            selfControl.attributes["eventtitle"].value = data[0].title;
                            selfControl.attributes["eventstart"].value = data[0].start;
                            selfControl.attributes["eventend"].value = data[0].end;
                            selfControl.attributes["eventallday"].value = data[0].allDay;
                            raiseEvent(selfControl.id, 'Control', 'OnChange');
                        },
                        error: function (text) {
                            alert('there was an error while trying to execute event drop');
                            revertFunc();
                        }
                    });

                }
                else
                    revertFunc();
            }
        });
    },

    _init: function () {
        
    },


    _setOption: function (option, value) {

        switch (option.toUpperCase()) {
            case "ISVISIBLE":
                this.isvisible(value);
                break;
            case "PROPERTY":
                this.property(value);
                break;
        }

        $.Widget.prototype._setOption.apply(this, arguments);
    },

    destroy: function () {
        $(this.element).unbind('.calendar1');

        $.Widget.prototype.destroy.call(this);
    },
    isvisible: function (value) {
        if (checkExists(value)) {
            if (value == false || value == 'false')
                this._calendar.css('display', 'none');
            else
                this._calendar.css('display', 'inline');
        }
        else {
            return !(this._calendar.css('display') == 'none');
        }

    },
    property: function (value) {
        if (checkExists(value)) {

        }
    },

    setStyles: function (styles) {
    }
});

})(jQuery);


$(document).ready(function () {
    //console.log("...SourceCodeANZ.Forms.Controls.Calendar...Loaded");
    jQuery('.SFC.SourceCodeANZ-Forms-Controls-Calendar').each(function (i, element) {

        var optionsValue = $(element).attr('data-options');
        var options = $.parseJSON(optionsValue);
        var viewsToShow = options.MonthView?"month,":"";
        viewsToShow += options.WeekView ? "agendaWeek," : "";
        viewsToShow += options.DayView ? "agendaDay," : "";
        viewsToShow += options.BasicWeekView ? "basicWeek," : "";
        viewsToShow += options.BasicDayView ? "basicDay," : "";
        viewsToShow += options.ListYearView ? "listYear," : "";
        viewsToShow += options.ListMonthView ? "listMonth," : "";
        viewsToShow += options.ListWeekView ? "listWeek," : "";
        viewsToShow += options.ListDayView ? "listDay," : "";
        if (viewsToShow.length == 0)
            viewsToShow = "month";
        $(element).attr('header_right', viewsToShow);
        var obj = $(element).calendar1(options);
    });

});

File: Source/Calendar.cs
using System;
using System.Collections.Generic;
using System.Xml;
using System.Text;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;
using SourceCode.Forms.Controls.Web.SDK;
using SourceCode.Forms.Controls.Web.SDK.Attributes;

[assembly: WebResource("SourceCodeANZ.Forms.Controls.Resources.Calendar.js", "text/javascript", PerformSubstitution = true)]
[assembly: WebResource("SourceCodeANZ.Forms.Controls.Resources.moment.min.js", "text/javascript", PerformSubstitution = true)]
[assembly: WebResource("SourceCodeANZ.Forms.Controls.Resources.fullcalendar.min.js", "text/javascript", PerformSubstitution = true)]
[assembly: WebResource("SourceCodeANZ.Forms.Controls.Resources.locale-all.js", "text/javascript", PerformSubstitution = true)]
[assembly: WebResource("SourceCodeANZ.Forms.Controls.Resources.fullcalendar.min.css", "text/css", PerformSubstitution = true)]
[assembly: WebResource("SourceCodeANZ.Forms.Controls.style.css", "text/css", PerformSubstitution = true)]
[assembly: WebResource("SourceCodeANZ.Forms.Controls.fullcal16.png", "image/png")]
[assembly: WebResource("SourceCodeANZ.Forms.Controls.Resources.CalendarPropertyConfig.js", "text/javascript", PerformSubstitution = true)]
namespace SourceCodeANZ.Forms.Controls
{
    [ControlTypeDefinition("SourceCodeANZ.Forms.Controls.Resources.CalendarDefinition.xml")]
    [ClientScript("SourceCodeANZ.Forms.Controls.Resources.Calendar.js")]
    [ClientScript("SourceCodeANZ.Forms.Controls.Resources.moment.min.js")]
    [ClientScript("SourceCodeANZ.Forms.Controls.Resources.fullcalendar.min.js")]
    [ClientScript("SourceCodeANZ.Forms.Controls.Resources.locale-all.js")]
    [ClientCss("SourceCodeANZ.Forms.Controls.Resources.fullcalendar.min.css")]
    class Calendar : BaseControl
    {

        private string _dataSourceType;

      
        public Calendar()
           : base("div")
        {
            ((SourceCode.Forms.Controls.Web.Shared.IControl)this).DesignFormattingPaths.Add("stlyecss", "SourceCodeANZ.Forms.Controls.style.css");
        }

        public string Value
        {
            get { return this.Attributes["value"]; }
            set { this.Attributes["value"] = value; }
        }


        public string EventTitle
        {
            get { return this.Attributes["eventtitle"]; }
            set { this.Attributes["eventtitle"] = value; }
        }

        public string FilterSpec
        {
            get { return this.Attributes["filterspec"]; }
            set { this.Attributes["filterspec"] = value; }
        }

        public string MethodParameters
        {
            get { return this.Attributes["methodparameters"]; }
            set { this.Attributes["methodparameters"] = value; }
        }

        public string SOCreateMethod
        {
            get { return this.Attributes["socreatemethod"]; }
            set { this.Attributes["socreatemethod"] = value; }
        }

        public string SOSaveMethod
        {
            get { return this.Attributes["sosavemethod"]; }
            set { this.Attributes["sosavemethod"] = value; }
        }


        public string EventTitleField
        {
            get { return this.GetOption<string>("EventTitleField", ""); }
            set { this.SetOption<string>("EventTitleField", value, ""); }
        }

               
        public string EventStart
        {
            get { return this.Attributes["eventstart"]; }
            set { this.Attributes["eventstart"] = value; }
        }


        public string EventStartField
        {
            get { return this.GetOption<string>("EventStartField", ""); }
            set { this.SetOption<string>("EventStartField", value, ""); }
        }

       
        public string EventEnd
        {
            get { return this.Attributes["eventend"]; }
            set { this.Attributes["eventend"] = value; }
        }


        public string EventEndField
        {
            get { return this.GetOption<string>("EventEndField", ""); }
            set { this.SetOption<string>("EventEndField", value, ""); }
        }

        
        public bool EventAllDay
        {
            get { return Convert.ToBoolean(this.Attributes["eventallday"]); }
            set { this.Attributes["eventallday"] = value.ToString(); }
        }


        public string EventAllDayField
        {
            get { return this.GetOption<string>("EventAllDayField", ""); }
            set { this.SetOption<string>("EventAllDayField", value, ""); }
        }

        public bool EventEditable
        {
            get { return Convert.ToBoolean(this.Attributes["eventeditable"]); }
            set { this.Attributes["eventeditable"] = value.ToString(); }
        }

        public string EventEditableField
        {
            get { return this.GetOption<string>("EventEditableField", ""); }
            set { this.SetOption<string>("EventEditableField", value, ""); }
        }

        public bool EventSizeable
        {
            get { return Convert.ToBoolean(this.Attributes["eventsizeable"]); }
            set { this.Attributes["eventsizeable"] = value.ToString(); }
        }

        public string EventSizeableField
        {
            get { return this.GetOption<string>("EventSizeableField", ""); }
            set { this.SetOption<string>("EventSizeableField", value, ""); }
        }

        public string EventColor
        {
            get { return this.Attributes["eventcolor"]; }
            set { this.Attributes["eventcolor"] = value; }
        }

        public string EventType
        {
            get { return this.Attributes["eventtype"]; }
            set { this.Attributes["eventtype"] = value; }
        }


        public string EventTypeField
        {
            get { return this.GetOption<string>("EventTypeField", ""); }
            set { this.SetOption<string>("EventTypeField", value, ""); }
        }


        public string EventColorField
        {
            get { return this.GetOption<string>("EventColorField", ""); }
            set { this.SetOption<string>("EventColorField", value, ""); }
        }


        public string EventDataSourceField
        {
            get { return this.GetOption<string>("EventDataSourceField", ""); }
            set { this.SetOption<string>("EventDataSourceField", value, ""); }
        }


        public string EventDataSourceIDfield
        {
            get { return this.GetOption<string>("EventDataSourceIDfield", ""); }
            set { this.SetOption<string>("EventDataSourceIDfield", value, ""); }
        }

        public Unit Width { get; set; }

        public string MonthViewColFormat
        {
            get { return this.Attributes["monthviewcolformat"]; }
            set { this.Attributes["monthviewcolformat"] = value; }
        }

        public string WeekViewColFormat
        {
            get { return this.Attributes["weekviewcolformat"]; }
            set { this.Attributes["weekviewcolformat"] = value; }
        }

        public string DayViewColFormat
        {
            get { return this.Attributes["dayviewcolformat"]; }
            set { this.Attributes["dayviewcolformat"] = value; }
        }

        public string MonthViewTitleFormat
        {
            get { return this.Attributes["monthviewtitleformat"]; }
            set { this.Attributes["monthviewtitleformat"] = value; }
        }

        public string WeekViewTitleFormat
        {
            get { return this.Attributes["weekviewtitleformat"]; }
            set { this.Attributes["weekviewtitleformat"] = value; }
        }

        public string DayViewTitleFormat
        {
            get { return this.Attributes["dayviewtitleformat"]; }
            set { this.Attributes["dayviewtitleformat"] = value; }
        }

        //public string HeaderLeft
        //{
        //    get { return this.Attributes["headerleft"]; }
        //    set { this.Attributes["headerleft"] = value; }
        //}

        //public string HeaderCenter
        //{
        //    get { return this.Attributes["headercenter"]; }
        //    set { this.Attributes["headercenter"] = value; }
        //}

        //public string HeaderRight
        //{
        //    get { return this.Attributes["headerright"]; }
        //    set { this.Attributes["headerright"] = value; }
        //}

        public bool CustomNewEventUI
        {
            get { return Convert.ToBoolean(this.Attributes["customneweventui"]); }
            set { this.Attributes["customneweventui"] = value.ToString(); }
        }

        public bool LoadDataOnInit
        {
            get { return Convert.ToBoolean(string.IsNullOrEmpty(this.Attributes["loaddataoninit"]) ? "True" : this.Attributes["loaddataoninit"]); }
            set { this.Attributes["loaddataoninit"] = value.ToString(); }
        }

        //UserServerZone
        public bool UseServerZone
        {
            get { return this.GetOption<bool>("UseServerZone", false); }
            set { this.SetOption<bool>("UseServerZone", value, false); }
        }

        public bool IsVisible
        {
            get { return this.GetOption<bool>("isvisible", true); }
            set { this.SetOption<bool>("isvisible", value, true); }
        }

        public bool IsEnabled
        {
            get { return this.GetOption<bool>("isenabled", true); }
            set { this.SetOption<bool>("isenabled", value, true); }
        }

        public bool IsReadOnly
        {
            get { return this.GetOption<bool>("isreadonly", false); }
            set { this.SetOption<bool>("isreadonly", value, false); }
        }

        public string AssociationMethod
        {
            get { return this.GetOption<string>("SmartObjectMethod", ""); }
            set { this.SetOption<string>("SmartObjectMethod", value, ""); }
        }

        public string AssociationSO
        {
            get { return this.GetOption<string>("SmartObject", ""); }
            set { this.SetOption<string>("SmartObject", value, ""); }
        }


        public string ValueProperty
        {
            get { return this.GetOption<string>("SMOValueProperty", ""); }
            set { this.SetOption<string>("SMOValueProperty", value, ""); }
        }

        public string DisplayTemplate
        {
            get { return this.GetOption<string>("SMODisplayTemplate", ""); }
            set { this.SetOption<string>("SMODisplayTemplate", value, ""); }
        }

        public string Locale
        {
            get { return this.GetOption<string>("Locale", ""); }
            set { this.SetOption<string>("Locale", value, ""); }
        }


        public string DataSourceType
        {
            get { return this._dataSourceType; }
            set { this._dataSourceType = value; }
        }

        public bool MonthView { get { return this.GetOption<bool>("MonthView", true); } set { this.SetOption<bool>("MonthView", value, true); } }
        public bool WeekView { get { return this.GetOption<bool>("WeekView", true); } set { this.SetOption<bool>("WeekView", value, true); } }
        public bool DayView { get { return this.GetOption<bool>("DayView", true); } set { this.SetOption<bool>("DayView", value, true); } }
        public bool BasicWeekView { get { return this.GetOption<bool>("BasicWeekView", false); } set { this.SetOption<bool>("BasicWeekView", value, false); } }
        public bool BasicDayView { get { return this.GetOption<bool>("BasicDayView", false); } set { this.SetOption<bool>("BasicDayView", value, false); } }
        public bool ListYearView { get { return this.GetOption<bool>("ListYearView", false); } set { this.SetOption<bool>("ListYearView", value, false); } }
        public bool ListMonthView { get { return this.GetOption<bool>("ListMonthView", false); } set { this.SetOption<bool>("ListMonthView", value, false); } }
        public bool ListWeekView { get { return this.GetOption<bool>("ListWeekView", true); } set { this.SetOption<bool>("ListWeekView", value, true); } }
        public bool ListDayView { get { return this.GetOption<bool>("ListDayView", false); } set { this.SetOption<bool>("ListDayView", value, false); } }



        protected override void CreateChildControls()
        {
            Attributes.CssStyle.Add("width", (!Width.IsEmpty) ? Width.ToString() : "100%");
            if (State == SourceCode.Forms.Controls.Web.Shared.ControlState.Designtime)
            {
                Panel pnlDevSurface = new Panel();
                pnlDevSurface.BorderStyle = BorderStyle.Solid;
                pnlDevSurface.BorderColor = System.Drawing.Color.RoyalBlue;
                pnlDevSurface.Height = Unit.Pixel(500);
                pnlDevSurface.Controls.Add(new LiteralControl("<p><strong>Desiger View:</strong><br/>The Actual Calendar will be rendered at runtime.</p>"));
                Controls.Add(pnlDevSurface);
            }
            Attributes.Add("value", (!string.IsNullOrEmpty(Value)) ? Value : string.Empty);
            Attributes.Add("eventtitle", (!string.IsNullOrEmpty(EventTitle)) ? EventTitle : string.Empty);
            Attributes.Add("filterspec", (!string.IsNullOrEmpty(FilterSpec)) ? FilterSpec : string.Empty);
            Attributes.Add("methodparameters", (!string.IsNullOrEmpty(MethodParameters)) ? MethodParameters : string.Empty);
            Attributes.Add("socreatemethod", (!string.IsNullOrEmpty(SOCreateMethod)) ? SOCreateMethod : "Create");
            Attributes.Add("sosavemethod", (!string.IsNullOrEmpty(SOSaveMethod)) ? SOSaveMethod : "Save");
            Attributes.Add("eventstart", (!string.IsNullOrEmpty(EventStart)) ? EventStart : string.Empty);
            Attributes.Add("eventend", (!string.IsNullOrEmpty(EventEnd)) ? EventEnd : string.Empty);
            Attributes.Add("eventallday", EventAllDay.ToString());
            Attributes.Add("eventeditable", EventEditable.ToString());
            Attributes.Add("eventsizeable", EventSizeable.ToString());
            Attributes.Add("loaddataoninit", LoadDataOnInit.ToString());
            Attributes.Add("eventcolor", (!string.IsNullOrEmpty(EventColor)) ? EventColor : string.Empty);
            Attributes.Add("eventtype", (!string.IsNullOrEmpty(EventType)) ? EventType : string.Empty);
            Attributes.Add("customneweventui", CustomNewEventUI.ToString());
            Attributes.Add("isreadonly", IsReadOnly.ToString().ToLower());

            Attributes.Add("monthviewcolformat", (!string.IsNullOrEmpty(MonthViewColFormat)) ? MonthViewColFormat : "ddd");
            Attributes.Add("weekviewcolformat", (!string.IsNullOrEmpty(WeekViewColFormat)) ? WeekViewColFormat : "ddd M/d");
            Attributes.Add("dayviewcolformat", (!string.IsNullOrEmpty(DayViewColFormat)) ? DayViewColFormat : "dddd M/d");

            Attributes.Add("monthviewtitleformat", (!string.IsNullOrEmpty(MonthViewTitleFormat)) ? MonthViewTitleFormat : "MMMM YYYY");
            Attributes.Add("weekviewtitleformat", (!string.IsNullOrEmpty(WeekViewTitleFormat)) ? WeekViewTitleFormat : "MMM D YYYY");
            Attributes.Add("dayviewtitleformat", (!string.IsNullOrEmpty(DayViewTitleFormat)) ? DayViewTitleFormat : "MMMM D YYYY");

            Attributes.CssStyle.Add("box-sizing", "border-box");
            Attributes.CssStyle.Add(HtmlTextWriterStyle.Display, (IsVisible||this.State == SourceCode.Forms.Controls.Web.Shared.ControlState.Designtime) ? "inline-block" : "none");
            
            if (DataSourceType == "Static")
            {
                Literal lit = new Literal();
                lit.Text = "Static Data as a Source is not Supported at this stage, Sorry!!";
                Controls.Add(lit);
            }

            if (!Options.ContainsKey("MonthView")) Options.Add("MonthView", true);
            if (!Options.ContainsKey("WeekView")) Options.Add("WeekView", true);
            if (!Options.ContainsKey("DayView")) Options.Add("DayView", true);
            if (!Options.ContainsKey("BasicWeekView")) Options.Add("BasicWeekView", false);
            if (!Options.ContainsKey("BasicDayView")) Options.Add("BasicDayView", false);
            if (!Options.ContainsKey("ListYearView")) Options.Add("ListYearView", false);
            if (!Options.ContainsKey("ListMonthView")) Options.Add("ListMonthView", false);
            if (!Options.ContainsKey("ListWeekView")) Options.Add("ListWeekView", true);
            if (!Options.ContainsKey("ListDayView")) Options.Add("ListDayView", false);
            if (!Options.ContainsKey("Locale")) Options.Add("Locale", "en");
            if (!Options.ContainsKey("UseServerZone")) Options.Add("UseServerZone", false);


            base.CreateChildControls();
            
         
        }

       
        

    }

    

}
